File: java/circular-buffer/src/main/java/org/mydotey/util/DataBuffer.java
package org.mydotey.util;

import java.util.Objects;
import java.util.function.Consumer;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public class DataBuffer<T> extends TimeSequenceCircularBuffer<DataBucket<T>> {

    public DataBuffer(DataBufferConfig bufferConfig) {
        super(bufferConfig);
    }

    @Override
    public DataBufferConfig getConfig() {
        return (DataBufferConfig) super.getConfig();
    }

    @Override
    protected DataBucket<T> newBucket(long startTime, long ttl) {
        return new DataBucket<T>(startTime, ttl, getConfig().getBucketCapacity());
    }

    public void add(T data) {
        getCurrentBucket().add(data);
    }

    public void consume(Consumer<T> consumer) {
        Objects.requireNonNull(consumer, "consumer is null");

        forEach(bucket -> {
            for (int i = 0; i < bucket.count(); i++) {
                T item = bucket.get(i);
                if (item == null)
                    continue;
                consumer.accept(item);
            }
        });
    }

}

File: dotnet/src/MyDotey.CircularBuffer/DataBuffer.cs
using System;

namespace MyDotey.Util
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public class DataBuffer<T> : TimeSequenceCircularBuffer<DataBucket<T>>
    {
        public DataBuffer(DataBufferConfig bufferConfig)
            : base(bufferConfig)
        {
        }

        public new DataBufferConfig Config
        {
            get { return (DataBufferConfig)base.Config; }
        }

        protected override DataBucket<T> NewBucket(long startTime, long ttl)
        {
            return new DataBucket<T>(startTime, ttl, Config.BucketCapacity);
        }

        public void Add(T data)
        {
            CurrentBucket.Add(data);
        }

        public void Consume(Action<T> consumer)
        {
            if (consumer == null)
                throw new ArgumentNullException("consumer is null");

            ForEach(bucket =>
            {
                for (int i = 0; i < bucket.Count; i++)
                {
                    T item = bucket.Get(i);
                    if (item == null)
                        continue;
                    consumer(item);
                }
            });
        }
    }
}